plugins/Utils/windowstatestorage.cpp

```diff
@@ -16,167 +17,319 @@
+                        Q_ARG(int, (int)windowState)
+                        );
+}
 
-    if (!query.first()) {
+WindowStateStorage::WindowState WindowStateStorage::getState(const QString &windowId, WindowStateStorage::WindowState defaultValue) const
+{
+    if (!m_asyncOk) {
         return defaultValue;
     }
+    int state;
 
-    const QRect result(query.value(QStringLiteral("x")).toInt(), query.value(QStringLiteral("y")).toInt(),
-                       query.value(QStringLiteral("width")).toInt(), query.value(QStringLiteral("height")).toInt());
+    QMetaObject::invokeMethod(m_asyncQuery, "getState", Qt::BlockingQueuedConnection,
+                            Q_RETURN_ARG(int, state),
+                            Q_ARG(const QString&, windowId)
+                            );
 
-    if (result.isValid()) {
-        return result;
+    if (state == -1) {
+        return defaultValue;
     }
 
-    return defaultValue;
+    return (WindowState)state;
 }
 
-void WindowStateStorage::initdb()
+int WindowStateStorage::getStage(const QString &appId, int defaultValue) const
 {
-    m_db.open();
-    if (!m_db.open()) {
-        qWarning() << "Error opening state database:" << m_db.lastError().driverText() << m_db.lastError().databaseText();
-        return;
+    if (!m_asyncOk) {
+        return defaultValue;
     }
+    int stage;
 
-    if (!m_db.tables().contains(QStringLiteral("geometry"))) {
-        QSqlQuery query;
-        query.exec(QStringLiteral("CREATE TABLE geometry(windowId TEXT UNIQUE, x INTEGER, y INTEGER, width INTEGER, height INTEGER);"));
-    }
+    QMetaObject::invokeMethod(m_asyncQuery, "getStage", Qt::BlockingQueuedConnection,
+                            Q_RETURN_ARG(int, stage),
+                            Q_ARG(const QString&, appId)
+                            );
 
-    if (!m_db.tables().contains(QStringLiteral("state"))) {
-        QSqlQuery query;
-        query.exec(QStringLiteral("CREATE TABLE state(windowId TEXT UNIQUE, state INTEGER);"));
+    if (stage == -1) {
+        return defaultValue;
     }
 
-    if (!m_db.tables().contains(QStringLiteral("stage"))) {
-        QSqlQuery query;
-        query.exec(QStringLiteral("CREATE TABLE stage(appId TEXT UNIQUE, stage INTEGER);"));
-    }
+    return stage;
 }
 
-void WindowStateStorage::saveValue(const QString &queryString)
+QRect WindowStateStorage::getGeometry(const QString &windowId, const QRect &defaultValue) const
 {
-    QMutexLocker mutexLocker(&s_mutex);
-    QSqlQuery query;
-    auto ok = query.exec(queryString);
-    if (!ok) {
-        qWarning() << "Error executing query" << queryString
-                   << "Driver error:" << query.lastError().driverText()
-                   << "Database error:" << query.lastError().databaseText();
+    if (!m_asyncOk) {
+        return defaultValue;
     }
+    QRect geometry;
+    QMetaObject::invokeMethod(m_asyncQuery, "getGeometry", Qt::BlockingQueuedConnection,
```

Comment:
I have a feeling that a read encountering a write in progress would freeze the UI due to the database being locked. Should we time out the queries to prevent that?

Reply:
I added a timeout in https://github.com/ubports/unity8/pull/356/commits/209b85132073915f87ba111155056e05d751ec89

Reply:
Wait... AFAIK, in SQLite a reader is never blocked; the only operation that can be blocked is the write one.

Reply:
In section 5.0 of [File Locking and Concurrency in SQLite Version 3](https://sqlite.org/lockingv3.html), it is specified that there is a short period of time during a write that readers are blocked. So it's totally possible, and it was cheap enough to add the timeout.